File: app/src/main/java/com/mytlogos/enterprise/ui/ViewerFragment.java
package com.mytlogos.enterprise.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.IdRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mytlogos.enterprise.R;
import com.mytlogos.enterprise.tools.ScrollHideHelper;
import com.omadahealth.github.swipyrefreshlayout.library.SwipyRefreshLayout;
import com.omadahealth.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

abstract class ViewerFragment<T> extends BaseFragment {
    private final ScrollHideHelper scrollHideHelper = new ScrollHideHelper();
    private View navigationView;
    private View appbar;
    private SwipyRefreshLayout swipeLayout;
    private TextView progressView;
    private View scrollView;
    private float progress = 0;
    private int maxScrolledY = 0;
    int currentEpisode;
    String currentBook;
    List<T> readableEpisodes = new ArrayList<>();
    T currentlyReading;
    static final String MEDIUM = "MEDIUM_FILE";
    static final String START_EPISODE = "START_EPISODE";

    @NonNull
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.viewer_layout, container, false);
        this.swipeLayout = view.findViewById(R.id.swiper);
        inflater.inflate(getLayoutRes(), this.swipeLayout, true);

        this.navigationView = view.findViewById(R.id.navigation);
        this.appbar = requireActivity().findViewById(R.id.appbar);
        this.swipeLayout.setOnRefreshListener(this::navigateEpisode);
        this.progressView = view.findViewById(R.id.progress);
        view.findViewById(R.id.left_nav).setOnClickListener(v -> navigateEpisode(SwipyRefreshLayoutDirection.TOP));
        view.findViewById(R.id.right_nav).setOnClickListener(v -> navigateEpisode(SwipyRefreshLayoutDirection.BOTTOM));

        int scrolledViewId = this.getScrolledViewId();
        if (scrolledViewId != View.NO_ID) {
            this.scrollView = view.findViewById(scrolledViewId);
            this.scrollView.setOnScrollChangeListener(
                    (v, scrollX, scrollY, oldScrollX, oldScrollY) ->
                            this.onScroll(scrollX, scrollY, oldScrollX, oldScrollY)
            );
        }
        return view;
    }

    void onScroll(int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
        if (scrollY != oldScrollY) {
            this.scrollHideHelper.hideGroups(oldScrollX, scrollX, oldScrollY, scrollY, this.navigationView, null, this.appbar, null);
        }
        if (scrollY > this.maxScrolledY) {
            this.maxScrolledY = scrollY;
            float progress = this.calculateProgressByScroll(scrollY, scrollX);
            this.updateProgress(progress);
        }
    }

    void toggleReadingMode() {
        this.scrollHideHelper.toggleGroups(this.navigationView, null, this.appbar, null);
    }

    void onLoadFinished() {
        this.maxScrolledY = 0;
        this.updateProgress(this.getCurrentProgress());
        this.seekFromProgress(this.progress);
        this.swipeLayout.setRefreshing(false);
    }

    @IdRes
    int getScrolledViewId() {
        return View.NO_ID;
    }

    void seekFromProgress(float progress) {

    }

    /**
     * Progress with value of 0 to 1.
     *
     * @param progress newProgress
     */
    void updateProgress(float progress) {
        if (progress > 1) {
            progress = 1;
        }
        progress = BigDecimal
                .valueOf(progress)
                .setScale(3, RoundingMode.CEILING)
                .floatValue();

        if (progress > this.progress) {
            this.progress = progress;
        } else {
            return;
        }
        progress = progress * 100;
        this.updateViewProgress(this.getProgressDescription(progress));
    }

    float calculateProgressByScroll(int scrollY, int scrollX) {
        if (this.scrollView == null || scrollY == 0) {
            return 0;
        }
        float maxHeight = 0;

        if (this.scrollView instanceof ViewGroup) {
            ViewGroup viewGroup = (ViewGroup) this.scrollView;

            for (int i = 0; i < viewGroup.getChildCount(); i++) {
                View child = viewGroup.getChildAt(i);
                maxHeight += child.getHeight();
            }
        } else {
            maxHeight = this.scrollView.getHeight();
        }
        scrollY = this.scrollView.getHeight() + scrollY;
        if (maxHeight == 0) {
            return 0;
        }
        return scrollY / maxHeight;
    }

    String getProgressDescription(float progress) {
        return String.format(Locale.getDefault(), "%.1f%%", progress);
    }

    void updateViewProgress(String progressDescription) {
        this.progressView.setText(progressDescription);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            this.currentEpisode = getArguments().getInt(START_EPISODE);
            this.currentBook = getArguments().getString(MEDIUM);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        this.scrollHideHelper.showGroups(this.navigationView, null, this.appbar, null);

        Bundle bundle = new Bundle();
        bundle.putInt(START_EPISODE, currentEpisode);
        bundle.putString(MEDIUM, currentBook);
        this.setArguments(bundle);
    }

    private void navigateEpisode(SwipyRefreshLayoutDirection direction) {
        if (this.currentlyReading == null) {
            if (this.readableEpisodes.isEmpty()) {
                return;
            } else {
                this.currentlyReading = this.readableEpisodes.get(0);
            }
        } else {
            int index = this.readableEpisodes.indexOf(currentlyReading);
            if (direction == SwipyRefreshLayoutDirection.TOP) {
                index--;
            } else if (direction == SwipyRefreshLayoutDirection.BOTTOM) {
                index++;
            } else {
                System.out.println("Unknown swipe direction in TextViewerFragment, neither top or bottom");
                return;
            }
            if (index >= this.readableEpisodes.size()) {
                // TODO: 26.07.2019 check with if there are more episodes and save them
                showToast("You are already reading the last saved episode");
                return;
            } else if (index < 0) {
                // TODO: 26.07.2019 check with if there are more episodes and save them
                showToast("You are already reading the first saved episode");
                return;
            }
            this.currentlyReading = this.readableEpisodes.get(index);
        }
        this.saveProgress(this.progress);
        this.updateContent();
    }

    @Override
    public void onStop() {
        super.onStop();
        this.saveProgress(this.progress);
    }

    abstract float getCurrentProgress();

    abstract void saveProgress(float progress);

    @LayoutRes
    abstract int getLayoutRes();

    abstract void updateContent();
}

File: app/src/main/java/com/mytlogos/enterprise/ui/ViewerFragment.kt
package com.mytlogos.enterprise.ui

import android.os.Bundle
import android.view.*
import android.widget.TextView
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import com.mytlogos.enterprise.R
import com.mytlogos.enterprise.tools.ScrollHideHelper
import com.omadahealth.github.swipyrefreshlayout.library.SwipyRefreshLayout
import com.omadahealth.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection
import com.omadahealth.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection.*
import kotlinx.coroutines.ExperimentalCoroutinesApi
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*

/**
 * An abstract ViewerFragment for Media.
 *
 * Flow:
 *
 * (ViewerFragment|SubClass):onCreateView
 *  - init views and listeners
 *
 * e.g. User scrolls
 *  -> onScroll gets called
 *
 * SubClass:onScroll
 * SubClass:seekFromProgress
 * SubClass:calculateProgressByScroll
 * SubClass:getProgressDescription
 * SubClass:updateViewProgress
 */
@ExperimentalCoroutinesApi
abstract class ViewerFragment<T> : BaseFragment() {
    private val scrollHideHelper = ScrollHideHelper()
    private lateinit var navigationView: View
    private lateinit var appbar: View
    private lateinit var swipeLayout: SwipyRefreshLayout
    private lateinit var progressView: TextView
    private var scrollView: View? = null
    private var progress = 0f
    private var maxScrolledY = 0

    var currentEpisode = 0
    var currentBook: String? = null
    var readableEpisodes: MutableList<T> = ArrayList()
    var currentlyReading: T? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.viewer_layout, container, false)
        swipeLayout = view.findViewById(R.id.swiper) as SwipyRefreshLayout
        inflater.inflate(layoutRes, swipeLayout, true)

        navigationView = view.findViewById(R.id.navigation)
        appbar = requireActivity().findViewById(R.id.appbar)

        swipeLayout.setOnRefreshListener(this@ViewerFragment::navigateEpisode)
        progressView = view.findViewById(R.id.progress) as TextView

        (view.findViewById(R.id.left_nav) as View).setOnClickListener { navigateEpisode(TOP) }
        (view.findViewById(R.id.right_nav) as View).setOnClickListener { navigateEpisode(BOTTOM) }

        if (scrolledViewId != View.NO_ID) {
            val localScrollView: View = view.findViewById(scrolledViewId)
            localScrollView.setOnScrollChangeListener { _: View?, scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int ->
                onScroll(scrollX,
                    scrollY,
                    oldScrollX,
                    oldScrollY
                )
            }
            scrollView = localScrollView
        }
        return view
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.viewer_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.read) {
            this.updateProgress(1.0f)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    open fun onScroll(scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int) {
        if (scrollY != oldScrollY) {
            scrollHideHelper.hideGroups(
                oldScrollX,
                scrollX,
                oldScrollY,
                scrollY,
                navigationView,
                null,
                appbar,
                null
            )
        }
        if (scrollY > maxScrolledY) {
            maxScrolledY = scrollY
            val progress = calculateProgressByScroll(scrollY, scrollX)
            updateProgress(progress)
        }
    }

    fun toggleReadingMode() {
        scrollHideHelper.toggleGroups(navigationView, null, appbar, null)
    }

    fun onLoadFinished() {
        maxScrolledY = 0
        updateProgress(currentProgress)
        seekFromProgress(progress)
        swipeLayout.isRefreshing = false
    }

    @get:IdRes
    open val scrolledViewId: Int
        get() = View.NO_ID

    open fun seekFromProgress(progress: Float) { /* no-op */ }

    /**
     * Progress with value of 0 to 1.
     *
     * @param progress newProgress
     */
    fun updateProgress(progress: Float) {
        var newProgress = progress
        if (newProgress > 1) {
            newProgress = 1f
        }
        newProgress = BigDecimal
            .valueOf(newProgress.toDouble())
            .setScale(3, RoundingMode.CEILING)
            .toFloat()
        if (newProgress > this.progress) {
            this.progress = newProgress
        } else {
            return
        }
        newProgress *= 100
        updateViewProgress(getProgressDescription(newProgress))
    }

    open fun calculateProgressByScroll(scrollY: Int, scrollX: Int): Float {
        var localScrollY = scrollY
        val localScrollView = scrollView
        if (localScrollView == null || localScrollY == 0) {
            return 0.0f
        }
        var maxHeight = 0f
        if (localScrollView is ViewGroup) {
            val viewGroup = localScrollView as ViewGroup?
            for (i in 0 until viewGroup!!.childCount) {
                val child = viewGroup.getChildAt(i)
                maxHeight += child.height.toFloat()
            }
        } else {
            maxHeight = localScrollView.height.toFloat()
        }
        localScrollY += localScrollView.height
        return if (maxHeight == 0f) {
            0.0f
        } else scrollY / maxHeight
    }

    open fun getProgressDescription(progress: Float): String {
        return String.format(Locale.getDefault(), "%.1f%%", progress)
    }

    open fun updateViewProgress(progressDescription: String?) {
        progressView.text = progressDescription
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            currentEpisode = requireArguments().getInt(START_EPISODE)
            currentBook = requireArguments().getString(MEDIUM)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        scrollHideHelper.showGroups(navigationView, null, appbar, null)
        val bundle = Bundle()
        bundle.putInt(START_EPISODE, currentEpisode)
        bundle.putString(MEDIUM, currentBook)
        this.arguments = bundle
    }

    private fun navigateEpisode(direction: SwipyRefreshLayoutDirection) {
        if (currentlyReading == null) {
            if (readableEpisodes.isEmpty()) {
                return
            } else {
                currentlyReading = readableEpisodes[0]
            }
        } else {
            var index = readableEpisodes.indexOf(currentlyReading)
            when (direction) {
                TOP -> index--
                BOTTOM -> index++
                else -> {
                    println("Unknown swipe direction in TextViewerFragment, neither top or bottom")
                    return
                }
            }
            if (index >= readableEpisodes.size) {
                // TODO: 26.07.2019 check with if there are more episodes and save them
                showToast("You are already reading the last saved episode")
                return
            } else if (index < 0) {
                // TODO: 26.07.2019 check with if there are more episodes and save them
                showToast("You are already reading the first saved episode")
                return
            }
            currentlyReading = readableEpisodes[index]
        }
        saveProgress(progress)
        updateContent()
    }

    override fun onStop() {
        super.onStop()
        saveProgress(progress)
    }

    abstract val currentProgress: Float
    abstract fun saveProgress(progress: Float)

    @get:LayoutRes
    abstract val layoutRes: Int
    abstract fun updateContent()

    companion object {
        const val MEDIUM = "MEDIUM_FILE"
        const val START_EPISODE = "START_EPISODE"
    }
}